### PR TITLE
Add Lua API for font configuration (#6710)

### LINKED ISSUE #6710: Lua function to change default font name and size
I am trying to write a simple plugin that has a few preset font configurations, like "heading 1", "heading 2", "body text", so I can quickly switch between these when making presentations.

However,  I don't see anything in [luapi_application.def.lua](https://github.com/xournalpp/xournalpp/blob/master/plugins/luapi_application.def.lua) that will allow me to change the default font name and size.

I do see `app.addTexts` which should add particular text strings, but that is not what I need. Just want to change the default font.

Is there a way to do this?

### PR DESCRIPTION
**Pull request by youssefadly237**
## Changes

- added `app.setFont()`,  `app.getFont`, `app.getFonts` to the Lua API
- added helper functions (`isFontFamilyAvailable`, `isFontDescriptionValid`) for validation

~**Note:** font validation was not implemented avoiding bloat~, implemented it anyway

## Testing

I have generated this with Claude 4.5 for testing, I hope I covered it all

```lua
local edgeCaseTestState = 0

function TestFontEdgeCasesCycle()
  local fonts = app.getFonts()

  -- Define edge case tests
  local tests = {
    -- Test 0: Very small font size
    function()
      print("=== Test 0: Very Small Font (0.1) ===")
      local success, err = pcall(function()
        app.setFont({ size = 0.1 })
      end)
      if success then
        print("ACCEPTED: Very small font size")
        local font = app.getFont()
        print("Result:", font.name, font.size)
      else
        print("REJECTED:", err)
      end
    end,

    -- Test 1: Zero font size (should fail)
    function()
      print("=== Test 1: Zero Font Size (Should Fail) ===")
      local success, err = pcall(function()
        app.setFont({ size = 0 })
      end)
      if not success then
        print("EXPECTED: Zero size rejected")
        print("Error:", err)
      else
        print("UNEXPECTED: Zero size accepted")
      end
    end,

    -- Test 2: Negative font size (should fail)
    function()
      print("=== Test 2: Negative Font Size (Should Fail) ===")
      local success, err = pcall(function()
        app.setFont({ size = -12 })
      end)
      if not success then
        print("EXPECTED: Negative size rejected")
        print("Error:", err)
      else
        print("UNEXPECTED: Negative size accepted")
      end
    end,

    -- Test 3: Very large font size (500)
    function()
      print("=== Test 3: Very Large Font (500) ===")
      local success, err = pcall(function()
        app.setFont({ size = 500 })
      end)
      if success then
        print("ACCEPTED: Large font size")
        local font = app.getFont()
        print("Result:", font.name, font.size)
      else
        print("REJECTED:", err)
      end
    end,

    -- Test 4: Even larger font size (1001)
    function()
      print("=== Test 4: Even Larger Font (1001) ===")
      local success, err = pcall(function()
        app.setFont({ size = 1001 })
      end)
      if success then
        print("ACCEPTED: Very large font size")
        local font = app.getFont()
        print("Result:", font.name, font.size)
      else
        print("REJECTED:", err)
      end
    end,

    -- Test 5: Extremely large font size (10000)
    function()
      print("=== Test 5: Extremely Large Font (10000) ===")
      local success, err = pcall(function()
        app.setFont({ size = 10000 })
      end)
      if success then
        print("ACCEPTED: Extremely large font size")
        local font = app.getFont()
        print("Result:", font.name, font.size)
      else
        print("REJECTED:", err)
      end
    end,

    -- Test 6: Infinity (should fail)
    function()
      print("=== Test 6: Infinity Font Size (Should Fail) ===")
      local success, err = pcall(function()
        app.setFont({ size = math.huge })
      end)
      if not success then
        print("EXPECTED: Infinity rejected")
        print("Error:", err)
      else
        print("UNEXPECTED: Infinity accepted")
      end
    end,

    -- Test 7: Negative infinity (should fail)
    function()
      print("=== Test 7: Negative Infinity (Should Fail) ===")
      local success, err = pcall(function()
        app.setFont({ size = -math.huge })
      end)
      if not success then
        print("EXPECTED: Negative infinity rejected")
        print("Error:", err)
      else
        print("UNEXPECTED: Negative infinity accepted")
      end
    end,

    -- Test 8: NaN (should fail)
    function()
      print("=== Test 8: NaN Font Size (Should Fail) ===")
      local success, err = pcall(function()
        app.setFont({ size = 0 / 0 })
      end)
      if not success then
        print("EXPECTED: NaN rejected")
        print("Error:", err)
      else
        print("UNEXPECTED: NaN accepted")
      end
    end,

    -- Test 9: Empty string font name (should fail)
    function()
      print("=== Test 9: Empty Font Name (Should Fail) ===")
      local success, err = pcall(function()
        app.setFont({ name = "" })
      end)
      if not success then
        print("EXPECTED: Empty name rejected")
        print("Error:", err)
      else
        print("UNEXPECTED: Empty name accepted")
      end
    end,

    -- Test 10: Whitespace-only font name (should fail)
    function()
      print("=== Test 10: Whitespace-Only Name (Should Fail) ===")
      local success, err = pcall(function()
        app.setFont({ name = "   " })
      end)
      if not success then
        print("EXPECTED: Whitespace name rejected")
        print("Error:", err)
      else
        print("ACCEPTED: Whitespace interpreted as valid font?")
        local font = app.getFont()
        print("Result:", font.name)
      end
    end,

    -- Test 11: Font name with special characters
    function()
      print("=== Test 11: Font Name With Numbers/Spaces ===")
      local specialFont = nil
      for _, family in ipairs(fonts.families) do
        if family:find("%d") or family:find(" ") then
          specialFont = family
          break
        end
      end
      if specialFont then
        local success, err = pcall(function()
          app.setFont({ name = specialFont, size = 12 })
        end)
        if success then
          print("ACCEPTED: Font with special characters:", specialFont)
          local font = app.getFont()
          print("Result:", font.name, font.size)
        else
          print("REJECTED:", err)
        end
      else
        print("No fonts with special characters found")
      end
    end,

    -- Test 12: Case sensitivity
    function()
      print("=== Test 12: Font Name Case Sensitivity ===")
      if #fonts.families > 0 then
        local originalName = fonts.families[1]
        local upperName = originalName:upper()
        print("Original:", originalName)
        print("Testing:", upperName)
        local success, err = pcall(function()
          app.setFont({ name = upperName, size = 12 })
        end)
        if success then
          print("ACCEPTED: Case-insensitive or uppercase font exists")
          local font = app.getFont()
          print("Actual font:", font.name)
        else
          print("REJECTED: Case-sensitive match required")
          print("Error:", err)
        end
      end
    end,

    -- Test 13: Invalid font family (should fail)
    function()
      print("=== Test 13: Invalid Font Family (Should Fail) ===")
      local success, err = pcall(function()
        app.setFont({ name = "NonExistentFont12345XYZ", size = 12 })
      end)
      if not success then
        print("EXPECTED: Invalid font rejected")
        print("Error:", err)
      else
        print("UNEXPECTED: Invalid font accepted")
      end
    end,

    -- Test 14: Table with wrong field names (should fail)
    function()
      print("=== Test 14: Wrong Field Names (Should Fail) ===")
      local success, err = pcall(function()
        app.setFont({ fontName = "Arial", fontSize = 12 })
      end)
      if not success then
        print("EXPECTED: Incorrect fields rejected")
        print("Error:", err)
      else
        print("UNEXPECTED: Incorrect fields accepted")
      end
    end,

    -- Test 15: Empty table (should fail)
    function()
      print("=== Test 15: Empty Table (Should Fail) ===")
      local success, err = pcall(function()
        app.setFont({})
      end)
      if not success then
        print("EXPECTED: Empty table rejected")
        print("Error:", err)
      else
        print("UNEXPECTED: Empty table accepted")
      end
    end,

    -- Test 16: Table with wrong types (should fail)
    function()
      print("=== Test 16: Wrong Types In Table (Should Fail) ===")
      local success, err = pcall(function()
        app.setFont({ name = 123, size = "large" })
      end)
      if not success then
        print("EXPECTED: Incorrect types rejected")
        print("Error:", err)
      else
        print("UNEXPECTED: Incorrect types accepted")
      end
    end,

    -- Test 17: Fractional font size
    function()
      print("=== Test 17: Fractional Font Size (12.5) ===")
      local success, err = pcall(function()
        app.setFont({ size = 12.5 })
      end)
      if success then
        print("ACCEPTED: Fractional size")
        local font = app.getFont()
        print("Result:", font.name, font.size)
      else
        print("REJECTED:", err)
      end
    end,

    -- Test 18: Pango string with size
    function()
      print("=== Test 18: Pango String Format ===")
      if #fonts.families > 0 then
        local testFont = fonts.families[1] .. " 16"
        print("Testing:", testFont)
        local success, err = pcall(function()
          app.setFont(testFont)
        end)
        if success then
          print("ACCEPTED: Pango string")
          local font = app.getFont()
          print("Result:", font.name, font.size)
        else
          print("REJECTED:", err)
        end
      end
    end,

    -- Test 19: Pango string without size (should fail)
    function()
      print("=== Test 19: Pango String Without Size (Should Fail) ===")
      if #fonts.families > 0 then
        local testFont = fonts.families[1]
        print("Testing:", testFont)
        local success, err = pcall(function()
          app.setFont(testFont)
        end)
        if not success then
          print("EXPECTED: Font without size rejected")
          print("Error:", err)
        else
          print("ACCEPTED: Possibly auto-sized font")
          local font = app.getFont()
          print("Result:", font.name, font.size)
        end
      end
    end,

    -- Test 20: Malformed Pango string (should fail)
    function()
      print("=== Test 20: Malformed Pango String (Should Fail) ===")
      local success, err = pcall(function()
        app.setFont("@#$%^&*() 12")
      end)
      if not success then
        print("EXPECTED: Malformed string rejected")
        print("Error:", err)
      else
        print("ACCEPTED: Possibly handled by Pango")
        local font = app.getFont()
        print("Result:", font.name, font.size)
      end
    end,

    -- Test 21: No arguments (should fail)
    function()
      print("=== Test 21: No Arguments (Should Fail) ===")
      local success, err = pcall(function()
        app.setFont()
      end)
      if not success then
        print("EXPECTED: Missing arguments rejected")
        print("Error:", err)
      else
        print("UNEXPECTED: No arguments accepted")
      end
    end,

    -- Test 22: Wrong argument type - number (should fail)
    function()
      print("=== Test 22: Number Argument (Should Fail) ===")
      local success, err = pcall(function()
        app.setFont(42)
      end)
      if not success then
        print("EXPECTED: Numeric argument rejected")
        print("Error:", err)
      else
        print("UNEXPECTED: Numeric argument accepted")
      end
    end,

    -- Test 23: Wrong argument type - boolean (should fail)
    function()
      print("=== Test 23: Boolean Argument (Should Fail) ===")
      local success, err = pcall(function()
        app.setFont(true)
      end)
      if not success then
        print("EXPECTED: Boolean argument rejected")
        print("Error:", err)
      else
        print("UNEXPECTED: Boolean argument accepted")
      end
    end,

    -- Test 24: Rapid font changes
    function()
      print("=== Test 24: Rapid Font Changes (10-30) ===")
      local success = true
      local count = 0
      for i = 10, 30 do
        local ok = pcall(function()
          app.setFont({ size = i })
        end)
        if ok then
          count = count + 1
        else
          success = false
          break
        end
      end
      if success then
        print("ACCEPTED: Handled", count, "rapid changes")
        local font = app.getFont()
        print("Final font:", font.name, font.size)
      else
        print("FAILED after", count, "changes")
      end
    end,

    -- Test 25: Unicode font names
    function()
      print("=== Test 25: Unicode Font Names ===")
      local unicodeFont = nil
      for _, family in ipairs(fonts.families) do
        if family:match("[^%z-\127]") then -- Non-ASCII
          unicodeFont = family
          break
        end
      end
      if unicodeFont then
        print("Found:", unicodeFont)
        local success, err = pcall(function()
          app.setFont({ name = unicodeFont, size = 12 })
        end)
        if success then
          print("ACCEPTED: Unicode font")
          local font = app.getFont()
          print("Result:", font.name, font.size)
        else
          print("REJECTED:", err)
        end
      else
        print("No Unicode font names found on this system")
      end
    end,

    -- Test 26: Multiple consecutive setFont calls
    function()
      print("=== Test 26: Multiple Consecutive Calls ===")
      local success = pcall(function()
        app.setFont({ size = 10 })
        app.setFont({ size = 20 })
        app.setFont({ size = 15 })
      end)
      if success then
        print("ACCEPTED: Multiple consecutive calls")
        local font = app.getFont()
        print("Final font:", font.name, font.size)
      else
        print("FAILED")
      end
    end,

    -- Test 27: Set font then get font
    function()
      print("=== Test 27: Set-Get Consistency Check ===")
      if #fonts.families > 0 then
        local testName = fonts.families[math.min(5, #fonts.families)]
        local testSize = 18
        print("Setting:", testName, testSize)

        local success = pcall(function()
          app.setFont({ name = testName, size = testSize })
        end)

        if success then
          local font = app.getFont()
          if font.name == testName and font.size == testSize then
            print("ACCEPTED: Set/Get values consistent")
          else
            print("INCONSISTENT VALUES:")
            print("  Expected:", testName, testSize)
            print("  Got:", font.name, font.size)
          end
        else
          print("FAILED: setFont operation failed")
        end
      end
    end,
  }

  -- Execute current test
  if edgeCaseTestState < #tests then
    edgeCaseTestState = edgeCaseTestState + 1
  else
    edgeCaseTestState = 1
  end

  print("\n" .. string.rep("=", 50))
  print("Edge Case Test:", edgeCaseTestState, "/", #tests)
  print(string.rep("=", 50))

  tests[edgeCaseTestState]()

  print(string.rep("=", 50) .. "\n")
end
```

Closes #6710

### CODE REVIEW

**Comment by abdullahkhalids** [comment]
Thanks so much for doing this.

Can I ask what happens if we pass an unknown font name? For example `name="asdfaf"`. It doesn't seem like there is any validation for it.

Also, is there is a way to get a list of possible font names? If this is possible within xournalpp, perhaps the documentation string should indicate how this can be done.

**Comment by youssefadly237** [comment]
> With `app.setFont` we may want to add `app.getFont`

I skipped `getFont` font because I dont think it is super useful, because you dont get to see the fonts just their names, so using the ui to choose the font made more sense for me, but it is not bad idea overall

> Possibly allowing to only change the font  name without the font size or the font size without the font name should be supported, e.g. `app.setFont({size = 14})` would only change the size.

yeah I will add this

> Can I ask what happens if we pass an unknown font name? For example name="asdfaf". It doesn't seem like there is any validation for it.

I did not implement validation to avoid bloating, also it does not cause any crashes, so just use with cation

> Also, is there is a way to get a list of possible font names? If this is possible within xournalpp, perhaps the documentation string should indicate how this can be done.

it is not yet implemented, next commits I will add it under `app.getFont`

thank you guys for your review, much appreciated

**Comment by rolandlo** [comment]
Please rebase on current master (since the merge of #6588 introduced some conflicts).

File: src/core/plugin/luapi_application.h [hunk 1]
```diff
@@ -3379,6 +3379,61 @@ static int applib_setPlaceholderValue(lua_State* L) {
...
+
+        newFont = XojFont(name, size);
+    } else {
+        return luaL_error(L, "Font must be either a string or a table with 'name' and 'size' fields");
+    }
+
+    // XojFont constructor handles format validation
+    if (newFont.getName().empty() || newFont.getSize() <= 0) {
+        return luaL_error(L, "Invalid font specification");
+    }
```
Comment (rolandlo) [inline src/core/plugin/luapi_application.h]: I think this only applies to the construction from a string description. So this should be moved to that case.

**Comment by rolandlo** [comment]
By the way in Lua using LuaGObject one can write
```lua
local LuaGObject = require "LuaGObject"
local Gtk = lgi.require("Gtk", "3.0")
local window = Gtk.Window {}
local context = window:create_pango_context()
local families = context:list_families()
for k,family in pairs(families) do print(k,family:get_name()) end
```
to print out the names of all font families installed on the system. It requires LuaGObject (or lgi) though.

Maybe `app.getFonts` could return a table containing a key `families` with an array of font family names and a key `current` with the index of the currently used font.

**Comment by rolandlo** [comment]
Fixes #6710

**Comment by rolandlo** [comment]
Thanks @youssefadly237 for this valuable contribution. I merged it.
